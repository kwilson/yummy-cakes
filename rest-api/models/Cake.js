class Cake {
    constructor(id, name, imageUrl, yumFactor, comment) {
        this.id = id;
        this.name = name;
        this.comment = comment;
        this.imageUrl = imageUrl;
        this.yumFactor = yumFactor;
    }
}

module.exports = Cake;