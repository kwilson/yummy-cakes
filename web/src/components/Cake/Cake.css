.cake {
    background: none;
    border: none;
    background-size: cover;
    background-position: 50% 50%;
    cursor: pointer;
    height: 100%;
    padding: 0;
    position: relative;
    width: 100%;
}

.cake__details {
    background-color: rgba(0,0,0,.4);
    bottom: 0;
    color: white;
    padding: 5px 10px;
    position: absolute;
    text-align: center;
    transition: background .25s linear;
    width: 100%;
}

.cake:hover .cake__details {
    background-color: rgba(0,0,0,.8);
}
