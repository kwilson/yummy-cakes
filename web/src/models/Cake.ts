export interface Cake {
  id: string;
  name: string;
  imageUrl: string;
  yumFactor: number;
  comment: string;
}
