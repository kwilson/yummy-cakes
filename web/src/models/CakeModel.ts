export interface CakeModel {
  id: string;
  name?: string;
  imageUrl?: string;
  yumFactor?: number;
  comment?: string;
}
