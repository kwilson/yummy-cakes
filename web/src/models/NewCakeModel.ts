export interface NewCakeModel {
  name?: string;
  imageUrl?: string;
  yumFactor?: number;
  comment?: string;
}
