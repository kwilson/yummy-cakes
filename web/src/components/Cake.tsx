import * as React from 'react';
import { CakeModel } from '../models/CakeModel';

import './Cake.css';

export interface CakeProps {
  cake: CakeModel;
  selectCake: () => any;
}

const Cake: React.StatelessComponent<CakeProps> = ({ cake, selectCake }) => {
  const onClick = () => selectCake();

  const style: React.CSSProperties = {
    backgroundImage: `url(${cake.imageUrl})`
  };

  const details = cake.name
    ? (
      <div className="cake__details">
        {cake.name}
      </div>
    )
    : null;

  return (
    <button className="cake" onClick={onClick} style={style}>
      {details}
    </button>
  );
};

export default Cake;
