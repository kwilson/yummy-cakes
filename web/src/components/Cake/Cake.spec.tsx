import * as React from 'react';
import { mount } from 'enzyme';
import { create as createSnapshot } from 'react-test-renderer';

import Cake from './Cake';
import { CakeModel } from '../../models/CakeModel';

describe('Cake Component', () => {
  let selectCakeHandler: jest.Mock<{}>;

  beforeEach(() => {
    selectCakeHandler = jest.fn();
  });

  it('renders an empty container if the model has no name', () => {
    // Arrange
    const cake: CakeModel = {
      id: 'ID123'
    };

    const cmp = <Cake cake={cake} selectCake={selectCakeHandler} />;

    // Act
    const el = mount(cmp);
    const snapshot = createSnapshot(cmp).toJSON();

    // Assert
    const details = el.find('.cake__details');
    expect(details).toHaveLength(0);

    expect(snapshot).toMatchSnapshot();
  });

  it('renders the name from the model', () => {
    // Arrange
    const cakeName = 'Name of the cake';
    const cake: CakeModel = {
      id: 'ID123',
      name: cakeName
    };

    const cmp = <Cake cake={cake} selectCake={selectCakeHandler} />;

    // Act
    const el = mount(cmp);
    const snapshot = createSnapshot(cmp).toJSON();

    // Assert
    const details = el.find('.cake__details');
    expect(details).toHaveLength(1);

    expect(details.text()).toEqual(cakeName);

    expect(snapshot).toMatchSnapshot();
  });

  it('triggers the select cake handler when the button is clicked', () => {
    // Arrange
    const cake: CakeModel = {
      id: 'ID123'
    };

    const cmp = <Cake cake={cake} selectCake={selectCakeHandler} />;

    // Act
    const el = mount(cmp);
    el.simulate('click');

    // Assert
    expect(selectCakeHandler).toHaveBeenCalledTimes(1);
  });
});
