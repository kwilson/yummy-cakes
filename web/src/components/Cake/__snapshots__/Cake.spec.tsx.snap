// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cake Component renders an empty container if the model has no name 1`] = `
<button
  className="cake"
  onClick={[Function]}
  style={
    Object {
      "backgroundImage": "url(undefined)",
    }
  }
/>
`;

exports[`Cake Component renders the name from the model 1`] = `
<button
  className="cake"
  onClick={[Function]}
  style={
    Object {
      "backgroundImage": "url(undefined)",
    }
  }
>
  <div
    className="cake__details"
  >
    Name of the cake
  </div>
</button>
`;
